src/app/ReadClient.cpp

```diff
@@ -894,6 +921,25 @@ void ReadClient::OnLivenessTimeoutCallback(System::Layer * apSystemLayer, void *
                  "Subscription Liveness timeout with SubscriptionID = 0x%08" PRIx32 ", Peer = %02x:" ChipLogFormatX64,
                  _this->mSubscriptionId, _this->GetFabricIndex(), ChipLogValueX64(_this->GetPeerNodeId()));
 
+    if (_this->mPeerActivePeriod != System::Clock::kZero)
+    {
+        // If device is sleeping, we mark the subscription as "IdleSubscription", and trigger resubscription on
+        // `OnActiveModeNotification`.
+        System::Clock::Timestamp time;
+        CHIP_ERROR error = System::SystemClock().GetClock_RealTimeMS(time);
+        if (error != CHIP_NO_ERROR)
+        {
+            ChipLogError(DataManagement, "Failed to get current time: %s", ErrorStr(error));
+            _this->Close(error, false);
+        }
+        if (time > _this->mPeerActiveUntilTimestamp)
```

Comment:
I don't understand why we need all this complexity.

Is the idea here that a subscription drop while the server is known to be active should lead to immediate resubscribe attempts, but a subscription drop when it's not known to be active should lead to just waiting for a check-in?

But in what case would we ever get a drop while the server is known to be active?  The MaxInterval is always going to be too long for that, no?

Reply:
> Is the idea here that a subscription drop while the server is known to be active should lead to immediate resubscribe attempts, but a subscription drop when it's not known to be active should lead to just waiting for a check-in?

Yes

> But in what case would we ever get a drop while the server is known to be active? The MaxInterval is always going to be too long for that, no?

As the active duration is a uint32, it is possible that the device may be active for a longer time and we just don't forbidden it.

In this case, failed to send a report (and we failed to resubscribe it) is something that we should notice the upper layer -- the device might be broken.

Reply:
But in practice all our ICD code sets the MaxInterval based on the gap between wakeup times, no?

Reply:
And in particular, an ICD that is active for longer than MaxInterval is supposed to send a checkin message, right?  Why wouldn't we just wait for that checkin message instead of guessing at whether we need to re-subscribe or not based on timing?

Reply:
Makes sense, updated.